crates/execution/src/iter.rs

```diff
@@ -165,96 +322,63 @@ impl<'a> Iterator for Iter<'a> {
-    /// If the lhs returns n-tuples,
-    /// this returns (n+1)-tuples.
-    MultiCol(MultiCol),
-    /// A single column index join.
-    /// If the lhs returns n-tuples,
-    /// this returns (n+1)-tuples.
-    Eq(SingleCol),
-}
-
-impl<'a, P, Q> Iterator for IxJoin<P, Q>
+/// An iterator for a left deep join tree
+pub enum LeftDeepJoin<Iter> {
+    /// A standard join
+    Eq(Iter),
+    /// A semijoin that returns the lhs
+    SemiLhs(Iter),
+    /// A semijion that returns the rhs
+    SemiRhs(Iter),
+}
+
+impl<'a, Iter> Iterator for LeftDeepJoin<Iter>
 where
-    P: Iterator<Item = (Tuple<'a>, RowRef<'a>)>,
-    Q: Iterator<Item = (Tuple<'a>, RowRef<'a>)>,
+    Iter: Iterator<Item = (Tuple<'a>, RowRef<'a>)>,
 {
     type Item = Tuple<'a>;
 
     fn next(&mut self) -> Option<Self::Item> {
-        let proj_left_deep_join = |(tuple, ptr)| {
-            match (tuple, ptr) {
-                // A leaf join
-                //   x
-                //  / \
-                // a   b
-                (Tuple::Row(u), ptr) => {
-                    // Returns a 2-tuple
-                    Tuple::Join(vec![u, Row::Ptr(ptr)])
-                }
-                // A left deep join
-                //     x
-                //    / \
-                //   x   c
-                //  / \
-                // a   b
-                (Tuple::Join(mut rows), ptr) => {
-                    // Returns an n+1 tuple
-                    rows.push(Row::Ptr(ptr));
-                    Tuple::Join(rows)
-                }
-            }
-        };
         match self {
             Self::SemiLhs(iter) => {
-                // A left semijoin
+                // Return the lhs tuple
                 iter.next().map(|(t, _)| t)
             }
             Self::SemiRhs(iter) => {
-                // A right semijoin
-                iter.next().map(|(_, ptr)| ptr).map(Row::Ptr).map(Tuple::Row)
-            }
-            Self::MultiColSemiLhs(iter) => {
-                // A left semijoin
-                iter.next().map(|(t, _)| t)
-            }
-            Self::MultiColSemiRhs(iter) => {
-                // A right semijoin
+                // Return the rhs row
                 iter.next().map(|(_, ptr)| ptr).map(Row::Ptr).map(Tuple::Row)
             }
-            Self::MultiCol(iter) => {
-                // Appends the rhs to the lhs
-                iter.next().map(proj_left_deep_join)
-            }
             Self::Eq(iter) => {
-                // Appends the rhs to the lhs
-                iter.next().map(proj_left_deep_join)
+                iter.next().map(|(tuple, ptr)| {
+                    match (tuple, ptr) {
+                        // A leaf join
+                        //   x
+                        //  / \
+                        // a   b
+                        (Tuple::Row(u), ptr) => {
+                            // Returns a 2-tuple
+                            Tuple::Join(vec![u, Row::Ptr(ptr)])
+                        }
```

Comment:
If we always return 2 then how about `[2; Row]`?

Reply:
This is for the leaves of the join